contracts/Staking.sol

```diff
@@ -0,0 +1,466 @@
+        success = true;
+      } else {
+        break;
+      }
+      idx++;
+    }
+    require(success, "rlp decode ack package failed");
+
+    require(ITokenHub(TOKEN_HUB_ADDR).withdrawStakingBNB(pack.amount), "withdraw funds from tokenhub failed");
+
+    delegated[pack.delegator] = delegated[pack.delegator].sub(pack.amount);
+    undelegated[pack.delegator] = undelegated[pack.delegator].add(pack.amount);
+    delegatedOfValidator[pack.delegator][pack.validator] = delegatedOfValidator[pack.delegator][pack.validator].sub(pack.amount);
+
+    emit failedDelegate(pack.delegator, pack.validator, pack.amount, pack.errCode);
+  }
+
+  function _handleUndelegateAckPackage(RLPDecode.Iterator memory iter) internal {
+    UndelegateAckPackage memory pack;
+
+    bool success = false;
+    uint256 idx = 0;
+    while (iter.hasNext()) {
+      if (idx == 0) {
+        pack.delegator = address(uint160(iter.next().toAddress()));
+      } else if (idx == 1) {
+        pack.validator = address(uint160(iter.next().toAddress()));
+      } else if (idx == 2) {
+        pack.amount = uint256(iter.next().toUint());
+      } else if (idx == 3) {
+        pack.errCode = uint8(iter.next().toUint());
+        success = true;
+      } else {
+        break;
+      }
+      idx++;
+    }
+    require(success, "rlp decode ack package failed");
+
+    delegated[pack.delegator] = delegated[pack.delegator].add(pack.amount);
+    pendingUndelegated[pack.delegator][pack.validator] = 0;
+
+    emit failedUndelegate(pack.delegator, pack.validator, pack.amount, pack.errCode);
+  }
+
+  function _handleRedelegateAckPackage(RLPDecode.Iterator memory iter) internal {
+    RedelegateAckPackage memory pack;
+
+    bool success = false;
+    uint256 idx = 0;
+    while (iter.hasNext()) {
+      if (idx == 0) {
+        pack.delegator = address(uint160(iter.next().toAddress()));
+      } else if (idx == 1) {
+        pack.valSrc = address(uint160(iter.next().toAddress()));
+      } else if (idx == 2) {
+        pack.valDst = address(uint160(iter.next().toAddress()));
+      } else if (idx == 3) {
+        pack.amount = uint256(iter.next().toUint());
+      } else if (idx == 4) {
+        pack.errCode = uint8(iter.next().toUint());
+        success = true;
+      } else {
+        break;
+      }
+      idx++;
+    }
+    require(success, "rlp decode ack package failed");
+
+    delegatedOfValidator[pack.delegator][pack.valSrc] = delegatedOfValidator[pack.delegator][pack.valSrc].add(pack.amount);
+    delegatedOfValidator[pack.delegator][pack.valDst] = delegatedOfValidator[pack.delegator][pack.valDst].sub(pack.amount);
+    pendingRedelegate[pack.delegator][pack.valSrc][pack.valDst] = 0;
+
+    emit failedRedelegate(pack.delegator, pack.valSrc, pack.valDst, pack.amount, pack.errCode);
+  }
+
+  function _handleDistributeRewardSynPackage(RLPDecode.Iterator memory iter) internal returns(uint32) {
+    DistributeRewardSynPackage memory pack;
```

Comment:
Do we have to define `DistributeRewardSynPackage` here? I think we can use `amount` and `recipient` as variable.

Reply:
It will be more readable